core/routing_dtlsr.go

```diff
@@ -0,0 +1,750 @@
+
+	dtlsr.dataMutex.RLock()
+	forwarder, present := dtlsr.routingTable[recipient]
+	dtlsr.dataMutex.RUnlock()
+	if !present {
+		// we don't know where to forward this bundle
+		log.WithFields(log.Fields{
+			"bundle":    bp.ID(),
+			"recipient": recipient,
+		}).Debug("DTLSR could not find a node to forward to")
+		return
+	}
+
+	for _, cs := range dtlsr.c.claManager.Sender() {
+		if cs.GetPeerEndpointID() == forwarder {
+			for _, eid := range sentEids {
+				if cs.GetPeerEndpointID() == eid {
+					log.WithFields(log.Fields{
+						"bundle": bndl.ID(),
+					}).Debug("Bundle was already forwarded to the appropriate peer")
+					return
+				}
+			}
+			sender = append(sender, cs)
+			log.WithFields(log.Fields{
+				"bundle":             bndl.ID(),
+				"recipient":          recipient,
+				"convergence-sender": sender,
+			}).Debug("DTLSR selected Convergence Sender for an outgoing bundle")
+			// we only ever forward to a single node
+
+			sentEids = append(sentEids, cs.GetPeerEndpointID())
+			bundleItem.Properties["routing/dtlsr/sent"] = sentEids
+			if err := dtlsr.c.store.Update(bundleItem); err != nil {
+				log.WithFields(log.Fields{
+					"error": err,
+				}).Warn("Updating BundleItem failed")
+			}
+
+			return
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"bundle":    bp.ID(),
+		"recipient": recipient,
+	}).Debug("DTLSR could not find forwarder amongst connected nodes")
+	return
+}
+
+func (dtlsr *DTLSR) ReportPeerAppeared(peer cla.Convergence) {
+	log.WithFields(log.Fields{
+		"address": peer,
+	}).Debug("Peer appeared")
+
+	peerReceiver, ok := peer.(cla.ConvergenceSender)
+	if !ok {
+		log.Warn("Peer was not a ConvergenceSender")
+		return
+	}
+
+	peerID := peerReceiver.GetPeerEndpointID()
+
+	log.WithFields(log.Fields{
+		"peer": peerID,
+	}).Debug("PeerID discovered")
+
+	dtlsr.dataMutex.Lock()
+	defer dtlsr.dataMutex.Unlock()
+	// track node
+	dtlsr.newNode(peerID)
+
+	// add node to peer list
+	dtlsr.peers.peers[peerID] = 0
+	dtlsr.peers.timestamp = timestampNow()
+	dtlsr.peerChange = true
+
+	log.WithFields(log.Fields{
+		"peer": peerID,
+	}).Debug("Peer is now being tracked")
+}
+
+func (dtlsr *DTLSR) ReportPeerDisappeared(peer cla.Convergence) {
+	log.WithFields(log.Fields{
+		"address": peer,
+	}).Debug("Peer disappeared")
+
+	peerReceiver, ok := peer.(cla.ConvergenceSender)
+	if !ok {
+		log.Warn("Peer was not a ConvergenceSender")
+		return
+	}
+
+	peerID := peerReceiver.GetPeerEndpointID()
+
+	log.WithFields(log.Fields{
+		"peer": peerID,
+	}).Debug("PeerID discovered")
+
+	dtlsr.dataMutex.Lock()
+	defer dtlsr.dataMutex.Unlock()
+	// set expiration timestamp for peer
+	timestamp := timestampNow()
+	dtlsr.peers.peers[peerID] = timestamp
+	dtlsr.peers.timestamp = timestamp
+	dtlsr.peerChange = true
+
+	log.WithFields(log.Fields{
+		"peer": peer,
+	}).Debug("Peer timeout is now running")
+}
+
+// DispatchingAllowed allows the processing of all packages.
+func (_ *DTLSR) DispatchingAllowed(_ BundlePack) bool {
```

Comment:
At this point, you could check whether this bundle has already been sent and whether the routing table has not changed in the meantime. This is not necessary, but would be a nice feature.

Reply:
Marked TODO for the future